sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/ParquetReadState.java

```diff
@@ -45,21 +47,41 @@
   /** Maximum definition level for the Parquet column */
   final int maxDefinitionLevel;
 
+  /** Maximum repetition level for the Parquet column */
+  final int maxRepetitionLevel;
+
   /** The current index over all rows within the column chunk. This is used to check if the
    * current row should be skipped by comparing against the row ranges. */
   long rowId;
 
-  /** The offset in the current batch to put the next value */
-  int offset;
+  /** The offset to put new values into definition & repetition level vector */
+  int levelOffset;
+
+  /** The offset to put new values into value vector */
+  int valueOffset;
 
   /** The remaining number of values to read in the current page */
   int valuesToReadInPage;
 
-  /** The remaining number of values to read in the current batch */
-  int valuesToReadInBatch;
+  /** The remaining number of rows to read in the current batch */
+  int rowsToReadInBatch;
+
+  // The following are only used when reading repeated values
+
+  /** When processing repeated values, whether we've found the beginning of the first list after the
+   *  current batch. */
```

Comment:
"after the current batch." or "in the current batch."?

Reply:
it is after the current batch, suppose we have the following repetition levels:
```
0, 1, 0, 1, 0, 1, 0, 1, 0, 1
```
and suppose batch size is 4, we'll need to see the 5th 0 (which is the first list of the _next_ batch) before we know that we've completed 4 lists for the batch.

Reply:
let me know if you have better way to write the comments, it could be confusing.